### ISSUE TITLE
Latex code directives produce inconsistent frames based on viewing resolution

### ISSUE THREAD

**Issue by AlexEshoo** [sphinx-doc/sphinx#2635]
When using the regular `::` code directive or the `code-block` directive, the generated formatting of the frame around the code is inconsistent a pdf viewer based on what level of zoom the user is set to. Is this due to the thickness of the frame?

See following images

76%
![76pct](https://cloud.githubusercontent.com/assets/8216513/15843455/4643819c-2c30-11e6-88cb-46fb61ec4c48.PNG)

100%
![100pct](https://cloud.githubusercontent.com/assets/8216513/15843456/4643a276-2c30-11e6-8ebd-f045f3dc3e71.PNG)

151%
![151pct](https://cloud.githubusercontent.com/assets/8216513/15843457/46493e70-2c30-11e6-89be-a2e673715e4b.PNG)

200%
![200pct](https://cloud.githubusercontent.com/assets/8216513/15843454/464329ae-2c30-11e6-8d63-1848fb27983c.PNG)

**Comment by andreacassioli** [comment]
Hi, I experience the same using `code-block` and `literal-include`. It seems viewer issue, but I still wondering whether it may also be elated with the package used for code blocks.

Related to #2450 ?

**Comment by jfbu** [comment]
This is a viewer + device issue, probably the same indeed as the one reported at #2450. Until recently I wasn't aware of it because on my mac os laptop I never see it, whether using Adobe Reader or Skim or Preview. On a desktop mac I again have no problem with Skim/Preview or whatever uses mac PDF libraries, but the issue you describes shows with Adobe Reader: the frames are partial until I zoom at 300%. This is the same software as on my mac laptop. I also checked that the artefact showed with a pdf as obtained using Sphinx 1.3.1, hence this is not a consequence of the recent refactoring done for code-blocks. Honestly, as I rarely use Adobe Reader, I did not give time to check if some of its options about anti-aliasing had an impact.

It is not a package problem, as down there the frame is drawn using the most basic TeX commands for rules. To be sure of that I will try to produce a minimal pure LaTeX example. It _may_ be a PDFTeX engine problem (or perhaps graphics pdftex driver), but the project I referred to above now uses LuaTeX and has the issue (anyway, LuaTeX possibly does as PDFTeX for such basic pdf code). I can also try to make a more minimal Sphinx project which I could compile using both make latex + dvipdfmx and make latexpdf and perhaps problem will show with the former but not latter moving the issue from pdf viewer to pdf producer.

I trust printing your document will be fine. I am very much interested however to know the source of this and implement a fix if at all possible on Sphinx side.

(much probably, increasing the thickness of frame rules should fix it too).

**Comment by jfbu** [comment]
From 

``` rst
This is a regular code directive::

  set stroke.index to 0.1mm

This is a code block directive.

.. code-block:: python

   set stroke.index to 0.1mm

1. This is a regular code directive in an enumerated list::

     set stroke index to 0.1mm

2. This is a code-block directive in an enumerated list.

   .. code-block:: python

      set stroke index to 0.1mm

```

and make latexpdf, then opening on Mac OS desktop with Skim at "real size", or any other zoom factor, I get correct

![capture d ecran 2016-06-07 a 09 40 39](https://cloud.githubusercontent.com/assets/2589111/15849852/dfbaacb0-2c94-11e6-9cb4-a918f05103a6.png)

However opening the same file in Adobe Reader, again at "real size", I get

![capture d ecran 2016-06-07 a 09 40 08](https://cloud.githubusercontent.com/assets/2589111/15849847/d71c1c56-2c94-11e6-9cd1-bc0b6902ff4f.png)

If I add dvipdfmx as document option, and produce via make latex, latex, dvipdfmx, Adobe will show

![capture d ecran 2016-06-07 a 09 42 56](https://cloud.githubusercontent.com/assets/2589111/15849874/ffaedf8c-2c94-11e6-815a-cdf2cd1e2e77.png)

which is deficient in another way. I get correct result at 300% zoom for both files.

I will check the Sphinx LaTeX code for perhaps a bug in the used thickness of rules but I don't think there is one.

**Comment by jfbu** [comment]
Can you check PR #2638. It seems to work on my device. But I will not have access to this desktop machine for some days, and can test on linux box only tomorrow.

The sphinx.sty controls completely the drawing of code frames if on single page, or the first part is split on pages. The rest uses standard `\fcolorbox` from package `xcolor`. Probably that package already applies the technique I added here to sphinx.sty which is to delay printing the frame until after the contents are drawn.

**Comment by AlexEshoo** [comment]
I tried that PR you made and it seems to work on my windows machine. I was unable to reproduce the output I was seeing before with any level of zoom. Seems to be a good fix.

**Comment by AlexEshoo** [comment]
Is there any way for me to renew this in the preamble of the document or am I stuck with editing sphinx.sty to solve this issue? The server I work on will be sticking with a particular version of sphinx for a while and I want to have a way to resolve some of these problems before an update. Since it is not a command, however, i assume I cannot use \renewcommand in my preamble to renew it.

**Comment by jfbu** [comment]
@AlexEshoo you can add this to your conf.py (sorry for length ;-) )

``` python
latex_elements = {
     # The paper size ('letterpaper' or 'a4paper').
     #
     # 'papersize': 'letterpaper',

     # The font size ('10pt', '11pt' or '12pt').
     #
     # 'pointsize': '10pt',

     # Additional stuff for the LaTeX preamble.
     #
     'preamble':
"""\
\\makeatletter
\\long\\def\\Sphinx@VerbatimFBox#1{%
  \\leavevmode
  \\begingroup
  % framed.sty does some measuring but this macro adds possibly a caption
  % use amsmath conditional to inhibit the caption counter stepping after
  % first pass
  \\ifSphinx@myfirstframedpass\\else\\firstchoice@false\\fi
  \\setbox\\@tempboxa\\hbox{\\kern\\fboxsep{#1}\\kern\\fboxsep}%
   \\hbox
   {\\lower\\dimexpr\\fboxrule+\\fboxsep+\\dp\\@tempboxa
    \\hbox{%
      \\vbox{\\ifx\\SphinxVerbatimTitle\\empty\\else
            % add the caption in a centered way above possibly indented frame
            % hide its width from framed.sty's measuring step
            % note that the caption brings \\abovecaptionskip top vertical space
            \\moveright\\dimexpr\\fboxrule+.5\\wd\\@tempboxa
                  \\hb@xt@\\z@{\\hss\\begin{minipage}{\\wd\\@tempboxa}%
                                  \\SphinxVerbatimTitle
                                 \\end{minipage}\\hss}\\fi
            % draw frame border _latest_ to avoid pdf viewer issue
            \\kern\\fboxrule
            \\hbox{\\kern\\fboxrule
                  \\vbox{\\vskip\\fboxsep\\copy\\@tempboxa\\vskip\\fboxsep}%
                  \\kern-\\wd\\@tempboxa\\kern-\\fboxrule
                  \\vrule\\@width\\fboxrule
                  \\kern\\wd\\@tempboxa
                  \\vrule\\@width\\fboxrule}%
            \\kern-\\dimexpr\\fboxsep+\\ht\\@tempboxa+\\dp\\@tempboxa
                                  +\\fboxsep+\\fboxrule\\relax
            \\hrule\\@height\\fboxrule
            \\kern\\dimexpr\\fboxsep+\\ht\\@tempboxa+\\dp\\@tempboxa+\\fboxsep\\relax
            \\hrule\\@height\\fboxrule}%
   }}%
  \\endgroup
  \\global\\Sphinx@myfirstframedpassfalse
}
\\makeatother
""",

     # further stuff
}
```

**Comment by jfbu** [comment]
In previous comment, I initially forgot the important `\makeatletter/\makeatother`.

**Comment by jfbu** [comment]
Let me comment more about underlying issue here which was already reported at #2450 (sorry @andreacassioli  for having thought it was not in Sphinx's reach). Using the following source

```
Hello
=====


.. raw:: latex

   \definecolor{VerbatimColor}{rgb}{0,0,1}
   \definecolor{VerbatimBorderColor}{rgb}{0,1,0}

This is a regular code directive::

  set stroke.index to 0.1mm

This is a code block directive.

.. code-block:: python

   set stroke.index to 0.1mm

1. This is a regular code directive in an enumerated list::

     set stroke index to 0.1mm

2. This is a code-block directive in an enumerated list.

   .. code-block:: python

      set stroke index to 0.1mm

```

in index.rst I have then made screenshots of a small section of a top frame, with either Acrobat Reader or with Skim (a PDF viewer for Mac OS X). This is on my laptop where actually Acrobat had no issues with frames, because I don't have access to the desktop now. Nevertheless there are obvious differences: here are some zoomed portions of three different screenshots.

With Skim:

![capture d ecran 2016-06-07 a 16 12 09](https://cloud.githubusercontent.com/assets/2589111/15866982/67922d60-2ce1-11e6-80cb-d666a8a1cd91.png)

With Acrobat Reader and stable branch of Sphinx

![capture d ecran 2016-06-07 a 16 09 35](https://cloud.githubusercontent.com/assets/2589111/15867000/79db9fa6-2ce1-11e6-818e-ad89867de7c8.png)

With Acrobat Reader and a pdf produced with  PR #2638.

![capture d ecran 2016-06-07 a 16 16 21](https://cloud.githubusercontent.com/assets/2589111/15867019/90db1fc4-2ce1-11e6-9b88-36bc4df2ce2b.png)

(for Skim it is as before PR). Sorry for varying dimensions. Observe how the green changes before and after the commit. It is darker in after the commit.

Obviously the PDF rendering by Acrobat does some things here which depend on the order  non-overlapping but contiguous areas are drawn. With black on white one can imagine this goes all the way to erasing.

**Comment by jfbu** [comment]
sorry the Skim screenshot was not large enough I was in a train and must have been fooled by some ambient light effet, here it is, more alike the two Adobe ones, 

![capture d ecran 2016-06-07 a 19 11 30](https://cloud.githubusercontent.com/assets/2589111/15867469/ca9b26f8-2ce3-11e6-8472-6d3d06dddacc.png)

there is a transition in the blue, but less pronounced as with Adobe. Anyway sorry for not very conclusive findings. Anyhow that's on the machine where I _don't_ see the problem ...

**Comment by jfbu** [comment]
Fix at af7d561 has been merged into stable, which will be released as 1.4.4. Thanks to @AlexEshoo  and @andreacassioli for reporting. Closing now.
